Missing types Errors

<!--
IF YOU DON'T FILL OUT THE FOLLOWING INFORMATION WE MIGHT CLOSE YOUR ISSUE WITHOUT INVESTIGATING
-->

**I'm submitting a ...**  
[x] bug report => search github for a similar issue or PR before submitting
[ ] feature request
[ ] support request


* **Angular version:** 5.2.X

#### Every time I run my project I have this errors caused by ng-formly

ERROR in node_modules/@ngx-formly/core/src/components/formly.field.config.d.ts(29,53): error TS7006: Parameter 'error' implicitly has an 'any' type.
node_modules/@ngx-formly/core/src/services/formly.config.d.ts(20,36): error TS7006: Parameter 'error' implicitly has an 'any' type.
node_modules/@ngx-formly/core/src/services/formly.config.d.ts(40,59): error TS7006: Parameter 'error' implicitly has an 'any' type.
node_modules/@ngx-formly/core/src/services/formly.config.d.ts(63,25): error TS7006: Parameter 'error' implicitly has an 'any' type.

--- a/.config/webpack.test.js
+++ b/.config/webpack.test.js
@@ -22,7 +22,7 @@ module.exports = {
       },
       {
         test: /\.ts$/,
-        use: ['awesome-typescript-loader?inlineSourceMap=true&sourceMap=false&declaration=false', 'angular2-template-loader'],
+        use: ['awesome-typescript-loader?inlineSourceMap=true&configFileName=tsconfig-spec.json&sourceMap=false&declaration=false', 'angular2-template-loader'],
         exclude: [/node_modules/]
       },
       { test: /\.html$/, use: ['raw-loader'] },

--- a/demo/src/tsconfig.app.json
+++ b/demo/src/tsconfig.app.json
@@ -1,5 +1,5 @@
 {
-  "extends": "../../tsconfig.json",
+  "extends": "../tsconfig.json",
   "compilerOptions": {
     "outDir": "../out-tsc/app",
     "module": "es2015"

--- a/demo/tsconfig.json
+++ b/demo/tsconfig.json
@@ -1,0 +1,10 @@
+{
+  "extends": "../tsconfig.json",
+  "compilerOptions": {
+    "noImplicitAny": false
+  },
+  "include": [
+    "../src/**/*.ts",
+    "src/**/*.ts"
+  ]
+}

--- a/package.json
+++ b/package.json
@@ -9,8 +9,8 @@
   "scripts": {
     "contrib:add": "all-contributors add",
     "contrib:generate": "all-contributors generate",
-    "lint": "tslint 'src/**/*.ts' 'demo/**/*.ts' -p tsconfig.json",
-    "lint:fix": "tslint --fix 'src/**/*.ts' 'app/**/*.ts' -p tsconfig.json",
+    "lint": "tslint -p tsconfig.json",
+    "lint:fix": "tslint --fix -p tsconfig.json",
     "postinstall": "patch --forward node_modules/ng-packagr/lib/steps/ngc.js .config/ng-packagr.patch || true",
     "commit": "npm run build && git-cz",
     "prepublish": "npm run build",

--- a/src/core/src/components/formly.attributes.ts
+++ b/src/core/src/components/formly.attributes.ts
@@ -55,7 +55,7 @@ export class FormlyAttributes implements OnChanges {
       return field.templateOptions[prop];
     }
 
-    return field[prop] || '';
+    return (<any>field)[prop] || '';
   }
 
   private getStatementValue(statement: string) {

--- a/src/core/src/components/formly.field.config.ts
+++ b/src/core/src/components/formly.field.config.ts
@@ -1,6 +1,7 @@
 import { FormGroup, AbstractControl, FormGroupDirective, NgForm } from '@angular/forms';
 import { Subject } from 'rxjs/Subject';
-import { Field } from './../templates/field';
+import { Field } from '../templates/field';
+import { TemplateManipulators } from '../services/formly.config';
 
 export interface FormlyFieldConfig {
   /**
@@ -32,7 +33,7 @@ export interface FormlyFieldConfig {
    */
   validation?: {
     messages?: {
-      [messageProperties: string]: string | ((error, field: FormlyFieldConfig) => string);
+      [messageProperties: string]: string | ((error: any, field: FormlyFieldConfig) => string);
     };
     show?: boolean;
     [additionalProperties: string]: any;
@@ -175,6 +176,7 @@ export interface FormlyTemplateOptions {
   click?: (field: FormlyFieldConfig, formControl: AbstractControl) => void;
   change?: (field: FormlyFieldConfig, formControl: AbstractControl) => void;
   keypress?: (field: FormlyFieldConfig, formControl: AbstractControl) => void;
+  templateManipulators?: TemplateManipulators;
   [additionalProperties: string]: any;
 }
 

--- a/src/core/src/components/formly.field.ts
+++ b/src/core/src/components/formly.field.ts
@@ -3,7 +3,7 @@ import {
   ViewContainerRef, ViewChild, ComponentRef, ComponentFactoryResolver, SimpleChanges,
 } from '@angular/core';
 import { FormGroup, FormArray } from '@angular/forms';
-import { FormlyConfig, TypeOption } from '../services/formly.config';
+import { FormlyConfig, TypeOption, TemplateManipulators } from '../services/formly.config';
 import { Field } from '../templates/field';
 import { FORMLY_VALIDATORS, evalExpression, getFieldModel } from '../utils';
 import { FormlyFieldConfig, FormlyFormOptions, FormlyValueChangeEvent } from './formly.field.config';
@@ -158,7 +158,7 @@ export class FormlyField implements OnInit, OnChanges, OnDestroy {
   }
 
   private getFieldWrappers(type: TypeOption) {
-    let templateManipulators = {
+    const templateManipulators: TemplateManipulators = {
       preWrapper: [],
       postWrapper: [],
     };
@@ -178,7 +178,7 @@ export class FormlyField implements OnInit, OnChanges, OnDestroy {
     return [...preWrappers, ...this.field.wrappers, ...postWrappers];
   }
 
-  private mergeTemplateManipulators(source, target) {
+  private mergeTemplateManipulators(source: TemplateManipulators, target: TemplateManipulators) {
     target = target || {};
     if (target.preWrapper) {
       source.preWrapper = source.preWrapper.concat(target.preWrapper);

--- a/src/core/src/services/formly.config.ts
+++ b/src/core/src/services/formly.config.ts
@@ -1,9 +1,9 @@
 import { Injectable, Inject, InjectionToken } from '@angular/core';
-import { ValidationErrors } from '@angular/forms';
+import { ValidationErrors, FormGroup } from '@angular/forms';
 import { FormlyGroup } from '../components/formly.group';
 import { Field } from './../templates/field';
 import { reverseDeepMerge } from './../utils';
-import { FormlyFieldConfig } from '../components/formly.field.config';
+import { FormlyFieldConfig, FormlyFormOptions } from '../components/formly.field.config';
 
 export const FORMLY_CONFIG_TOKEN = new InjectionToken<FormlyConfig>('FORMLY_CONFIG_TOKEN');
 
@@ -20,7 +20,7 @@ export class FormlyConfig {
   };
   validators: { [name: string]: ValidatorOption } = {};
   wrappers: { [name: string]: WrapperOption } = {};
-  messages: { [name: string]: string | ((error, field: FormlyFieldConfig) => string); } = {};
+  messages: { [name: string]: string | ((error: any, field: FormlyFieldConfig) => string); } = {};
 
   templateManipulators: {
     preWrapper: ManipulatorWrapper[];
@@ -31,7 +31,7 @@ export class FormlyConfig {
   };
 
   extras: {
-    fieldTransform?: any,
+    fieldTransform?: ((fields: FormlyFieldConfig[], model: any, form: FormGroup, options: FormlyFormOptions) => FormlyFieldConfig[])[],
     showError?: (field: Field) => boolean;
   } = {
     fieldTransform: undefined,
@@ -165,7 +165,7 @@ export class FormlyConfig {
     return this.validators[name];
   }
 
-  addValidatorMessage(name: string, message: string | ((error, field: FormlyFieldConfig) => string)) {
+  addValidatorMessage(name: string, message: string | ((error: any, field: FormlyFieldConfig) => string)) {
     this.messages[name] = message;
   }
 
@@ -213,7 +213,7 @@ export interface ValidatorOption {
 
 export interface ValidationMessageOption {
   name: string;
-  message: string | ((error, field: FormlyFieldConfig) => string);
+  message: string | ((error: any, field: FormlyFieldConfig) => string);
 }
 
 export interface ManipulatorOption {
@@ -223,6 +223,11 @@ export interface ManipulatorOption {
 
 export interface ManipulatorWrapper {
   (f: FormlyFieldConfig): string;
+}
+
+export interface TemplateManipulators {
+  preWrapper?: ManipulatorWrapper[];
+  postWrapper?: ManipulatorWrapper[];
 }
 
 export interface ConfigOption {

--- a/src/core/src/services/formly.form.builder.ts
+++ b/src/core/src/services/formly.form.builder.ts
@@ -111,7 +111,7 @@ export class FormlyFormBuilder {
     });
   }
 
-  private initFieldExpression(field: FormlyFieldConfig, model, options: FormlyFormOptions) {
+  private initFieldExpression(field: FormlyFieldConfig, model: any, options: FormlyFormOptions) {
     if (field.expressionProperties) {
       for (let key in field.expressionProperties as any) {
         if (typeof field.expressionProperties[key] === 'string' || isFunction(field.expressionProperties[key])) {
@@ -153,7 +153,7 @@ export class FormlyFormBuilder {
   }
 
   private initFieldAsyncValidation(field: FormlyFieldConfig) {
-    let validators = [];
+    let validators: any = [];
     if (field.asyncValidators) {
       for (let validatorName in field.asyncValidators) {
         if (validatorName !== 'validation') {
@@ -164,7 +164,7 @@ export class FormlyFormBuilder {
             }
 
             return new Promise((resolve) => {
-              return validator(control, field).then(result => {
+              return validator(control, field).then((result: boolean) => {
                 resolve(result ? null : {[validatorName]: true});
               });
             });
@@ -173,7 +173,7 @@ export class FormlyFormBuilder {
       }
     }
     if (field.asyncValidators && Array.isArray(field.asyncValidators.validation)) {
-      field.asyncValidators.validation.forEach(validate => {
+      field.asyncValidators.validation.forEach((validate: any) => {
         if (typeof validate === 'string') {
           validators.push(this.formlyConfig.getValidator(validate).validation);
         } else {
@@ -194,7 +194,7 @@ export class FormlyFormBuilder {
   }
 
   private initFieldValidation(field: FormlyFieldConfig) {
-    let validators = [];
+    let validators: any = [];
     FORMLY_VALIDATORS
       .filter(opt => (field.templateOptions && field.templateOptions[opt])
         || (field.expressionProperties && field.expressionProperties[`templateOptions.${opt}`]),
@@ -225,7 +225,7 @@ export class FormlyFormBuilder {
     }
 
     if (field.validators && Array.isArray(field.validators.validation)) {
-      field.validators.validation.forEach(validate => {
+      field.validators.validation.forEach((validate: any) => {
         if (typeof validate === 'string') {
           validators.push(this.formlyConfig.getValidator(validate).validation);
         } else {
@@ -287,7 +287,7 @@ export class FormlyFormBuilder {
     }
   }
 
-  private getValidation(opt, value) {
+  private getValidation(opt: string, value: any) {
     switch (opt) {
       case 'required':
         return Validators.required;

--- a/src/core/src/utils.ts
+++ b/src/core/src/utils.ts
@@ -9,8 +9,8 @@ export function getFieldId(formId: string, options: FormlyFieldConfig, index: st
 
 export function getKeyPath(field: {key?: string|string[], fieldGroup?: any, fieldArray?: any}): (string|number)[] {
   /* We store the keyPath in the field for performance reasons. This function will be called frequently. */
-  if (field['_formlyKeyPath'] !== undefined) {
-    return field['_formlyKeyPath'];
+  if ((<any> field)['_formlyKeyPath'] !== undefined) {
+    return (<any> field)['_formlyKeyPath'];
   }
   let keyPath: (string|number)[] = [];
   if (field.key) {
@@ -32,7 +32,7 @@ export function getKeyPath(field: {key?: string|string[], fieldGroup?: any, fiel
       }
     }
   }
-  field['_formlyKeyPath'] = keyPath;
+  (<any> field)['_formlyKeyPath'] = keyPath;
   return keyPath;
 }
 
@@ -66,7 +66,7 @@ export function getFieldModel(model: any, field: FormlyFieldConfig, constructEmp
   return value;
 }
 
-export function assignModelValue(model, path, value) {
+export function assignModelValue(model: any, path: string | (string | number)[], value: any) {
   if (typeof path === 'string') {
     path = getKeyPath({key: path});
   }
@@ -74,7 +74,7 @@ export function assignModelValue(model, path, value) {
   if (path.length > 1) {
     const e = path.shift();
     if (!model[e] || !isObject(model[e])) {
-      model[e] = isNaN(path[0]) ? {} : [];
+      model[e] = typeof path[0] === 'string' ? {} : [];
     }
     assignModelValue(model[e], path, value);
   } else {
@@ -82,14 +82,14 @@ export function assignModelValue(model, path, value) {
   }
 }
 
-export function getValueForKey(model, path) {
+export function getValueForKey(model: any, path: string | (string | number)[]): any {
   if (typeof path === 'string') {
     path = getKeyPath({key: path});
   }
   if (path.length > 1) {
     const e = path.shift();
     if (!model[e]) {
-      model[e] = isNaN(path[0]) ? {} : [];
+      model[e] = typeof path[0] === 'string' ? {} : [];
     }
     return getValueForKey(model[e], path);
   } else {
@@ -101,7 +101,7 @@ export function getKey(controlKey: string, actualKey: string) {
   return actualKey ? actualKey + '.' + controlKey : controlKey;
 }
 
-export function reverseDeepMerge(dest, ...args) {
+export function reverseDeepMerge(dest: any, ...args: any[]) {
   args.forEach(src => {
     for (let srcArg in src) {
       if (isNullOrUndefined(dest[srcArg]) || isBlankString(dest[srcArg])) {
@@ -134,7 +134,7 @@ export function isFunction(value: any) {
   return typeof(value) === 'function';
 }
 
-export function objAndSameType(obj1, obj2) {
+export function objAndSameType(obj1: any, obj2: any) {
   return isObject(obj1) && isObject(obj2) &&
     Object.getPrototypeOf(obj1) === Object.getPrototypeOf(obj2);
 }

--- a/src/ui-bootstrap/src/types/select.ts
+++ b/src/ui-bootstrap/src/types/select.ts
@@ -6,6 +6,7 @@ export class SelectOption {
   value?: string;
   group?: SelectOption[];
   disabled?: boolean;
+  [key: string]: any;
 
   constructor(label: string, value?: string, children?: SelectOption[]) {
     this.label = label;

--- a/src/ui-material/src/types/checkbox.ts
+++ b/src/ui-material/src/types/checkbox.ts
@@ -25,7 +25,7 @@ export class FormlyFieldCheckbox extends FieldType implements AfterViewInit {
   }
 
   ngAfterViewInit() {
-    const formField = this.field['__formField__'];
+    const formField = (<any> this.field)['__formField__'];
     if (formField) {
       formField._control.focusMonitor([this.matCheckbox._inputElement.nativeElement]);
 

--- a/src/ui-material/src/types/input.ts
+++ b/src/ui-material/src/types/input.ts
@@ -23,8 +23,8 @@ export class FormlyFieldInput extends FieldType implements OnInit {
   }
 
   ngOnInit() {
-    if (this.field['__formField__']) {
-      this.field['__formField__']._control = this.matInput;
+    if ((<any> this.field)['__formField__']) {
+      (<any> this.field)['__formField__']._control = this.matInput;
     }
     super.ngOnInit();
   }

--- a/src/ui-material/src/types/multicheckbox.ts
+++ b/src/ui-material/src/types/multicheckbox.ts
@@ -32,7 +32,7 @@ export class FormlyFieldMultiCheckbox extends FieldType implements AfterViewInit
   }
 
   ngAfterViewInit() {
-    const formField = this.field['__formField__'];
+    const formField = (<any> this.field)['__formField__'];
     if (formField) {
       formField._control.focusMonitor(
         this.matCheckboxes.map(matCheckbox => matCheckbox._inputElement.nativeElement),

--- a/src/ui-material/src/types/radio.ts
+++ b/src/ui-material/src/types/radio.ts
@@ -19,7 +19,7 @@ export class FormlyFieldRadio extends FieldType implements AfterViewInit {
   }
 
   ngAfterViewInit() {
-    const formField = this.field['__formField__'];
+    const formField = (<any>this.field)['__formField__'];
     if (formField) {
       formField._control.focusMonitor(
         this.matRadioButtons.map(matRadioButton => matRadioButton._inputElement.nativeElement),

--- a/src/ui-material/src/types/select.ts
+++ b/src/ui-material/src/types/select.ts
@@ -8,6 +8,7 @@ export class SelectOption {
   value?: string;
   group?: SelectOption[];
   disabled?: boolean;
+  [key: string]: any;
 
   constructor(label: string, value?: string, children?: SelectOption[]) {
     this.label = label;
@@ -76,8 +77,8 @@ export class FormlyFieldSelect extends FieldType implements OnInit {
   }
 
   ngOnInit() {
-    if (this.field['__formField__']) {
-      this.field['__formField__']._control = this.matSelect;
+    if ((<any>this.field)['__formField__']) {
+      (<any> this.field)['__formField__']._control = this.matSelect;
     }
     super.ngOnInit();
   }

--- a/src/ui-material/src/wrappers/form-field.ts
+++ b/src/ui-material/src/wrappers/form-field.ts
@@ -66,19 +66,19 @@ export class FormlyWrapperFormField extends FieldWrapper implements OnInit, OnDe
   ngOnInit() {
     this.focused = !!this.field.focus;
     this.formField._control = this;
-    this.field['__formField__'] = this.formField;
+    (<any> this.field)['__formField__'] = this.formField;
     if (this.to.floatPlaceholder) {
       this.to.floatLabel = this.to.floatPlaceholder;
       console.warn(`${this.field.key}: Passing 'floatPlaceholder' is deprecated, Use 'floatLabel' instead.`);
     }
   }
 
-  focusMonitor(elements = []) {
+  focusMonitor(elements: any[] = []) {
     elements.map(element => {
       takeUntil.call(
         this._focusMonitor.monitor(element, this.renderer, false),
         this.destroy$,
-      ).subscribe(focusOrigin => {
+      ).subscribe((focusOrigin: any) => {
         if (this.focused !== !!focusOrigin) {
           this.ngZone.run(() => {
             this.focused = !!focusOrigin;
@@ -93,7 +93,7 @@ export class FormlyWrapperFormField extends FieldWrapper implements OnInit, OnDe
   onContainerClick() {}
 
   ngOnDestroy() {
-    delete this.field['__formField__'];
+    delete (<any> this.field)['__formField__'];
     this.stateChanges.complete();
     this.destroy$.complete();
   }

--- a/tsconfig-spec.json
+++ b/tsconfig-spec.json
@@ -1,0 +1,10 @@
+{
+  "extends": "./tsconfig.json",
+  "compilerOptions": {
+    "noImplicitAny": false
+  },
+  "include": [
+    "src/**/*.ts",
+    "**/*.spec.ts"
+  ]
+}

--- a/tsconfig.json
+++ b/tsconfig.json
@@ -8,16 +8,16 @@
     "moduleResolution": "node",
     "removeComments": true,
     "sourceMap": true,
-    "noImplicitAny": false,
+    "noImplicitAny": true,
     "noUnusedLocals": false,
     "baseUrl": "src",
     "lib": ["dom", "es6"],
     "importHelpers": true,
     "paths": {
       // fix for #605
-      // TODO: Remove it after upgrading Angular to v.5      
+      // TODO: Remove it after upgrading Angular to v.5
       "@angular/*": ["../node_modules/@angular/*"],
-      
+
       "@ngx-formly/core": ["./core/index"],
       "@ngx-formly/bootstrap": ["./ui-bootstrap/index"],
       "@ngx-formly/material": ["./ui-material/index"]
@@ -27,8 +27,11 @@
     ]
   },
   "include": [
-    "src/**/*.ts",
-    "demo/**/*.ts"
+    "src/**/*.ts"
+  ],
+  "exclude": [
+    "test.ts",
+    "**/*.spec.ts"
   ],
   "compileOnSave": false
 }

